tests/Unit/Domain/CoordinateMaps/Test_CylindricalEndcap.cpp

```diff
@@ -0,0 +1,218 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#include "Framework/TestingFramework.hpp"
+
+#include <array>
+#include <boost/optional.hpp>
+#include <cmath>
+#include <gsl/gsl_poly.h>
+#include <random>
+
+#include "Domain/CoordinateMaps/CylindricalEndcap.hpp"
+#include "Helpers/Domain/CoordinateMaps/TestMapHelpers.hpp"
+#include "NumericalAlgorithms/RootFinding/QuadraticEquation.hpp"
+
+namespace domain {
+namespace {
+void test_cylindrical_endcap() {
+  INFO("CylindricalEndcap");
+  // Set up random number generator
+  MAKE_GENERATOR(gen);
+
+  std::uniform_real_distribution<> unit_dis(0.0, 1.0);
+  std::uniform_real_distribution<> interval_dis(-1.0, 1.0);
+  std::uniform_real_distribution<> angle_dis(0.0, 2.0 * M_PI);
+
+  // Choose some random centers for sphere_one and sphere_two
+  const std::array<double, 3> center_one = {
+      interval_dis(gen), interval_dis(gen), interval_dis(gen)};
+  CAPTURE_PRECISE(center_one);
+  const std::array<double, 3> center_two = {
+      interval_dis(gen), interval_dis(gen), interval_dis(gen)};
+  CAPTURE_PRECISE(center_two);
+  const double dist_between_spheres =
+      sqrt(square(center_two[0] - center_one[0]) +
+           square(center_two[1] - center_one[1]) +
+           square(center_two[2] - center_one[2]));
+
+  // Pick radius of sphere_one not too small compared to the distance
+  // between the centers.
+  const double radius_one = 0.3 * dist_between_spheres + unit_dis(gen);
+  CAPTURE_PRECISE(radius_one);
+
+  // Make sure z_plane intersects sphere_one on the +z side of the
+  // center. We don't allow the plane to be displaced by less than 10%
+  // or more than 90% of the radius.
+  const double z_plane =
+      center_one[2] + (0.1 + 0.8 * unit_dis(gen)) * radius_one;
+  CAPTURE_PRECISE(z_plane);
+
+  // Now construct sphere_two which we make sure encloses sphere_one,
+  // but doesn't have too large of a radius.
+  const double radius_two =
+      (unit_dis(gen) + 1.0) * (radius_one + dist_between_spheres);
+  CAPTURE_PRECISE(radius_two);
+
+  const std::array<double, 3> proj_center = [
+    &z_plane, &center_one, &radius_one, &center_two, &radius_two, &gen,
+    &unit_dis, &angle_dis
+  ]() noexcept {
+    // Consider a cone formed by taking the intersection of z_plane and
+    // sphere_one (this is a circle called circle_one) and connecting it to
+    // center_one. Compute theta, where 2*theta is the opening angle of this
+    // cone. Call this cone 'cone_one'.
+    const double cos_theta = (z_plane - center_one[2]) / radius_one;
+
+    // We will construct two new cones. The first we will call
+    // cone_regular. It is constructed so that cone_regular and
+    // cone_one intersect each other on circle_one at right angles.
```

Comment:
A figure here would be helpful...maybe refer to a figure in the hpp files somewhere, if an existing figure is suitable?